core/src/toga/images.py

```diff
@@ -1,53 +1,86 @@
 from __future__ import annotations
 
+from typing import Any
+
 from pathlib import Path
 
 import toga
 from toga.platform import get_platform_factory
 
+from io import BytesIO, BufferedReader
+import time, os
 
+try:
+    from PIL import Image as PIL_Image
+    PIL_ImportError_Message = None
+except ImportError as e:
+    PIL_Image = None
+    PIL_ImportError_Message = e
 class Image:
     def __init__(
         self,
-        path: str | None | Path = None,
+        src:str|Path|BytesIO|BufferedReader|bytes|PIL_Image.Image|None=None,
         *,
+        path: str | None | Path = None,
         data: bytes | None = None,
     ):
         """Create a new image.
 
-        An image must be provided either a ``path`` or ``data``, but not both.
-
+        :param src: string path, pathlib.Path, BufferedReader, BytesIO, bytes, Pillow Object of the image to load
         :param path: Path to the image to load. This can be specified as a string, or as
             a :any:`pathlib.Path` object. The path can be an absolute file system path,
             or a path relative to the module that defines your Toga application class.
         :param data: A bytes object with the contents of an image in a supported format.
         :raises FileNotFoundError: If a path is provided, but that path does not exist.
         :raises ValueError: If the path or data cannot be loaded as an image.
         """
-        if path is None and data is None:
-            raise ValueError("Either path or data must be set.")
-        if path is not None and data is not None:
-            raise ValueError("Only either path or data can be set.")
-
+        none_count = [src, path, data].count(None)
+        if none_count != 2:
+            raise ValueError("One and Only one of the three parameters (src, path, data) have to be set.")
+        
+        if src is not None:
+            if isinstance(src, str) or isinstance(src, Path):
+                path = src
+                src = None
+            elif isinstance(src, BytesIO) or isinstance(src, BufferedReader):
+                src.seek(0)
+                data = src.read()
+                src = None
+            elif isinstance(src, bytes):
+                data = src
+                src = None
+            elif PIL_Image is not None and isinstance(src, PIL_Image.Image):
+                img_buffer = BytesIO()
+                src.save(img_buffer, format=src.format)
+                img_buffer.seek(0)
+                data = img_buffer.read()
+                src = None
+            else:
+                raise TypeError("Unsupported source type for image.")
+                
         if path is not None:
             if isinstance(path, Path):
                 self.path = path
             else:
                 self.path = Path(path)
             self.data = None
-        else:
-            self.path = None
+        if data is not None:
```

Comment:
I'm fairly certain this leaves open a path `toga.Image()` won't raise an error. The test suite *should* catch this... but the test suite isn't currently running, because of other problems with the test suite.